// ngx-typed-js/bundles/ngx-typed-js.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('typed.js')) :
    typeof define === 'function' && define.amd ? define('ngx-typed-js', ['exports', '@angular/core', 'typed.js'], factory) :
    (global = global || self, factory(global['ngx-typed-js'] = {}, global.ng.core, global.Typed));
}(this, (function (exports, core, Typed) { 'use strict';

    Typed = Typed && Object.prototype.hasOwnProperty.call(Typed, 'default') ? Typed['default'] : Typed;

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __exportStar(m, exports) {
        for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var NgxTypedJsComponent = /** @class */ (function () {
        function NgxTypedJsComponent() {
            this.completed = new core.EventEmitter();
            this.preStringTyped = new core.EventEmitter();
            this.stringTyped = new core.EventEmitter();
            this.lastStringBackspaced = new core.EventEmitter();
            this.typingPaused = new core.EventEmitter();
            this.typingResumed = new core.EventEmitter();
            this.reset = new core.EventEmitter();
            this.stopped = new core.EventEmitter();
            this.started = new core.EventEmitter();
            this.destroyed = new core.EventEmitter();
        }
        NgxTypedJsComponent.prototype.ngAfterViewInit = function () {
            this.typed = new Typed(this.content.nativeElement.querySelector('.typing'), this.options);
            if (this.showCursor !== false) {
                this.updateCursorStyle();
            }
        };
        NgxTypedJsComponent.prototype.toggle = function () {
            this.typed.toggle();
        };
        NgxTypedJsComponent.prototype.stop = function () {
            this.typed.stop();
        };
        NgxTypedJsComponent.prototype.start = function () {
            this.typed.start();
        };
        NgxTypedJsComponent.prototype.destroy = function () {
            this.typed.destroy();
        };
        NgxTypedJsComponent.prototype.doReset = function (restart) {
            this.typed.reset(restart);
        };
        Object.defineProperty(NgxTypedJsComponent.prototype, "options", {
            get: function () {
                var emit = function (emitter) { return function () { return emitter.emit(); }; };
                var emitIndex = function (emitter) { return function (index) { return emitter.emit(index); }; };
                var opts = {
                    strings: this.strings,
                    stringsElement: this.stringsElement,
                    typeSpeed: this.typeSpeed,
                    startDelay: this.startDelay,
                    backSpeed: this.backSpeed,
                    smartBackspace: this.smartBackspace,
                    shuffle: this.shuffle,
                    backDelay: this.backDelay,
                    fadeOut: this.fadeOut,
                    fadeOutClass: this.fadeOutClass,
                    fadeOutDelay: this.fadeOutDelay,
                    loop: this.loop,
                    loopCount: this.loopCount,
                    showCursor: this.showCursor,
                    autoInsertCss: this.autoInsertCss,
                    attr: this.attr,
                    bindInputFocusEvents: this.bindInputFocusEvents,
                    contentType: this.contentType,
                    onComplete: emit(this.completed),
                    preStringTyped: emitIndex(this.preStringTyped),
                    onStringTyped: emitIndex(this.stringTyped),
                    onLastStringBackspaced: emit(this.lastStringBackspaced),
                    onTypingPaused: emitIndex(this.typingPaused),
                    onTypingResumed: emitIndex(this.typingResumed),
                    onReset: emit(this.reset),
                    onStop: emitIndex(this.stopped),
                    onStart: emitIndex(this.started),
                    onDestroy: emit(this.destroyed),
                };
                Object.keys(opts).forEach(function (key) {
                    if (opts[key] === undefined) {
                        delete opts[key];
                    }
                });
                return opts;
            },
            enumerable: true,
            configurable: true
        });
        NgxTypedJsComponent.prototype.updateCursorStyle = function () {
            var textElementStyle = getComputedStyle(this.content.nativeElement.querySelector('.typing'));
            var cursorElementStyle = this.content.nativeElement.querySelector('.typed-cursor').style;
            cursorElementStyle.fontSize = textElementStyle.fontSize;
            cursorElementStyle.color = this.cursorColor || textElementStyle.color;
        };
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxTypedJsComponent.prototype, "cursorColor", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Array)
        ], NgxTypedJsComponent.prototype, "strings", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxTypedJsComponent.prototype, "stringsElement", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Number)
        ], NgxTypedJsComponent.prototype, "typeSpeed", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Number)
        ], NgxTypedJsComponent.prototype, "startDelay", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Number)
        ], NgxTypedJsComponent.prototype, "backSpeed", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Boolean)
        ], NgxTypedJsComponent.prototype, "smartBackspace", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Boolean)
        ], NgxTypedJsComponent.prototype, "shuffle", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Number)
        ], NgxTypedJsComponent.prototype, "backDelay", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Boolean)
        ], NgxTypedJsComponent.prototype, "fadeOut", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxTypedJsComponent.prototype, "fadeOutClass", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Boolean)
        ], NgxTypedJsComponent.prototype, "fadeOutDelay", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Boolean)
        ], NgxTypedJsComponent.prototype, "loop", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Number)
        ], NgxTypedJsComponent.prototype, "loopCount", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Boolean)
        ], NgxTypedJsComponent.prototype, "showCursor", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxTypedJsComponent.prototype, "cursorChar", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Boolean)
        ], NgxTypedJsComponent.prototype, "autoInsertCss", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxTypedJsComponent.prototype, "attr", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Boolean)
        ], NgxTypedJsComponent.prototype, "bindInputFocusEvents", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxTypedJsComponent.prototype, "contentType", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", core.EventEmitter)
        ], NgxTypedJsComponent.prototype, "completed", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", core.EventEmitter)
        ], NgxTypedJsComponent.prototype, "preStringTyped", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", core.EventEmitter)
        ], NgxTypedJsComponent.prototype, "stringTyped", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", core.EventEmitter)
        ], NgxTypedJsComponent.prototype, "lastStringBackspaced", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", core.EventEmitter)
        ], NgxTypedJsComponent.prototype, "typingPaused", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", core.EventEmitter)
        ], NgxTypedJsComponent.prototype, "typingResumed", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", core.EventEmitter)
        ], NgxTypedJsComponent.prototype, "reset", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", core.EventEmitter)
        ], NgxTypedJsComponent.prototype, "stopped", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", core.EventEmitter)
        ], NgxTypedJsComponent.prototype, "started", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", core.EventEmitter)
        ], NgxTypedJsComponent.prototype, "destroyed", void 0);
        __decorate([
            core.ViewChild('wrapper', { static: true }),
            __metadata("design:type", Object)
        ], NgxTypedJsComponent.prototype, "content", void 0);
        NgxTypedJsComponent = __decorate([
            core.Component({
                selector: 'ngx-typed-js',
                template: "<div #wrapper>\n  <ng-content></ng-content>\n</div>\n",
                styles: [":host>>>.typing{display:inline}"]
            })
        ], NgxTypedJsComponent);
        return NgxTypedJsComponent;
    }());

    var NgxTypedJsModule = /** @class */ (function () {
        function NgxTypedJsModule() {
        }
        NgxTypedJsModule = __decorate([
            core.NgModule({
                declarations: [NgxTypedJsComponent],
                imports: [],
                exports: [NgxTypedJsComponent]
            })
        ], NgxTypedJsModule);
        return NgxTypedJsModule;
    }());

    exports.NgxTypedJsComponent = NgxTypedJsComponent;
    exports.NgxTypedJsModule = NgxTypedJsModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
